src/lib/protocols/connections/ConnectionService.ts

```diff
@@ -68,7 +68,8 @@ class ConnectionService extends EventEmitter {
     });
 
     connectionRecord.invitation = invitation;
-    this.connectionRepository.update(connectionRecord);
```

Comment:
As a side note, we should not make this state changed event emitting manual. We should attach the event emitter to the the repository/record/something that if the state of a record is changed, the event will automatically be emitted, because this is asking for inconsistencies.

ACA-Py has implemented this is in the base record. All records that extend this base record class can automatically emit state changed events (by sending webhooks).

Reply:
I agree. The way ACA-Py implemented this makes sense because they're using a ProfileSession. However, because we are currently not using such a mechanism, implementing this in `BaseRecord ` would mean we'd have to redundantly pass the corresponding `EventEmitter` class to the method that updates the state.

Emitting events from the `Repository` class therefore sounds more appealing because all _state-changing-records_ are passed to the `save()` or `update()` methods at some point. Here we could emit an event for every `save()` and `update()` call where there is a state change. However, the `Repository` class doesn't know about the `Record`s previous state. ACA-Py solves this by adding a `_last_state` property to `BaseRecord` so that might be an approach.

Another (possibly less significant) limitation of this approach is that  **if** one decides to use an event listener from within the framework, we probably have to make `Repository` a singleton. If we don't and multiple instances of let's say `Repository<ConnectionRecord>` are created, it would also create two separate `EventEmitter`s which would get messy very quickly. This would also require some refactoring since `Repository` is generic and I don't think making a generic class a singleton is a good idea.

The most viable approach I can think of at this moment is instead of letting the `Service` or `Repository` classes extend `EventEmitter`, create a separate (singleton) EventEmitter class that handles the events for all record classes and expose it to the framework's public API. This way we could also emit events from the `save()` and `update()` in `Repository` while eliminating the need to refactor.

The only issue with this _grand-central-event-emitter_ approach I currently see (I'm sure there are more) is that it could result in a boatload of event types. The `EventEmitter` API only allows you to filter based on a event name `emitter.on(eventName, listener)` and if we handle all event types for all records in a single `EventEmitter` instance, over time, we'd have a lot of event names. However, for the time being it might be a hacky solution.

I make these comments with the assumption that we're not moving towards a ACA-Py like `ProfileSession` based architecture anytime soon. If that assumption is false, then forget what I said.

Any thoughts?